File: client-scalaz/src/main/scala/org/openapitools/client/api/PurchaseItemFullResponse.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime


import PurchaseItemFullResponse._

case class PurchaseItemFullResponse (
  id: Option[Long],
name: Option[String],
description: Option[String],
tickets: Option[Integer],
coverAsset: Option[AssetShortResponse],
promoAsset: Option[AssetShortResponse],
giftable: Option[Boolean],
assetable: Option[Boolean],
gameType: Option[String],
price: Option[Float],
purchaseType: Option[String],
purchaseCode: Option[String],
purchaseLimit: Option[Integer],
quantity: Option[Integer],
serviceAction: Option[ServiceAction],
appName: Option[String],
secretKey: Option[String],
appKey: Option[String],
allocateTickets: Option[Boolean],
ticketType: Option[String],
ticketCount: Option[Long],
points: Option[Long])

object PurchaseItemFullResponse {
  import DateTimeCodecs._
  sealed trait ServiceAction
  case object DAYPREMIUM extends ServiceAction
  case object WEEKPREMIUM extends ServiceAction
  case object MONTHPREMIUM extends ServiceAction
  case object YEARPREMIUM extends ServiceAction
  case object LIFETIMEPREMIUM extends ServiceAction
  case object ADDTICKET extends ServiceAction
  case object ADDGIFT extends ServiceAction
  case object ADDPOINTS extends ServiceAction

  object ServiceAction {
    def toServiceAction(s: String): Option[ServiceAction] = s match {
      case "DAYPREMIUM" => Some(DAYPREMIUM)
      case "WEEKPREMIUM" => Some(WEEKPREMIUM)
      case "MONTHPREMIUM" => Some(MONTHPREMIUM)
      case "YEARPREMIUM" => Some(YEARPREMIUM)
      case "LIFETIMEPREMIUM" => Some(LIFETIMEPREMIUM)
      case "ADDTICKET" => Some(ADDTICKET)
      case "ADDGIFT" => Some(ADDGIFT)
      case "ADDPOINTS" => Some(ADDPOINTS)
      case _ => None
    }

    def fromServiceAction(x: ServiceAction): String = x match {
      case DAYPREMIUM => "DAYPREMIUM"
      case WEEKPREMIUM => "WEEKPREMIUM"
      case MONTHPREMIUM => "MONTHPREMIUM"
      case YEARPREMIUM => "YEARPREMIUM"
      case LIFETIMEPREMIUM => "LIFETIMEPREMIUM"
      case ADDTICKET => "ADDTICKET"
      case ADDGIFT => "ADDGIFT"
      case ADDPOINTS => "ADDPOINTS"
    }
  }

  implicit val ServiceActionEnumEncoder: EncodeJson[ServiceAction] =
    EncodeJson[ServiceAction](is => StringEncodeJson(ServiceAction.fromServiceAction(is)))

  implicit val ServiceActionEnumDecoder: DecodeJson[ServiceAction] =
    DecodeJson.optionDecoder[ServiceAction](n => n.string.flatMap(jStr => ServiceAction.toServiceAction(jStr)), "ServiceAction failed to de-serialize")

  implicit val PurchaseItemFullResponseCodecJson: CodecJson[PurchaseItemFullResponse] = CodecJson.derive[PurchaseItemFullResponse]
  implicit val PurchaseItemFullResponseDecoder: EntityDecoder[PurchaseItemFullResponse] = jsonOf[PurchaseItemFullResponse]
  implicit val PurchaseItemFullResponseEncoder: EntityEncoder[PurchaseItemFullResponse] = jsonEncoderOf[PurchaseItemFullResponse]
}

File: client-clojure/src/sirqul_io_t_platform/specs/purchase_item_full_response.clj
(ns sirqul-io-t-platform.specs.purchase-item-full-response
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.asset-short-response :refer :all]
            [sirqul-io-t-platform.specs.asset-short-response :refer :all]
            )
  (:import (java.io File)))


(def purchase-item-full-response-data
  {
   (ds/opt :id) int?
   (ds/opt :name) string?
   (ds/opt :description) string?
   (ds/opt :tickets) int?
   (ds/opt :coverAsset) asset-short-response-spec
   (ds/opt :promoAsset) asset-short-response-spec
   (ds/opt :giftable) boolean?
   (ds/opt :assetable) boolean?
   (ds/opt :gameType) string?
   (ds/opt :price) float?
   (ds/opt :purchaseType) string?
   (ds/opt :purchaseCode) string?
   (ds/opt :purchaseLimit) int?
   (ds/opt :quantity) int?
   (ds/opt :serviceAction) string?
   (ds/opt :appName) string?
   (ds/opt :secretKey) string?
   (ds/opt :appKey) string?
   (ds/opt :allocateTickets) boolean?
   (ds/opt :ticketType) string?
   (ds/opt :ticketCount) int?
   (ds/opt :points) int?
   })

(def purchase-item-full-response-spec
  (ds/spec
    {:name ::purchase-item-full-response
     :spec purchase-item-full-response-data}))
